docs/manual/src/internals/async.md

```diff
@@ -0,0 +1,159 @@
+  Rust-->>-Foreign: return RustFuture handle
+
+  loop until RustFuture is ready
+    Foreign->>Rust: call RustFuture poll fn
+    Rust->>Foreign: call future callback
+  end
+
+  Foreign->>+Rust: call RustFuture complete fn
+  Rust-->>-Foreign: return result
+  Foreign->>Rust: call RustFuture free fn
+  Foreign-->>-User: return from async function
+```
+
+For each async function, UniFFI generates 4 scaffolding functions:
+* A scaffolding function that returns a `RustFuture` handle
+* `rust_future_poll` to poll the future
+* `rust_future_complete` to receive the completed result of a future
+* `rust_future_free` to destroy a future freeing any underlying memory
+* Technically, the function names are more like `uniffi_[module_prefix]_rust_future_poll`, but this document uses above names make for shorthand.
+
+The UniFFI generated code for an async function performs these steps:
+
+1. Call the Rust scaffolding function, receiving a `RustFuture` handle
+1. Call the `rust_future_poll` function for the future until the future is ready.
+   * The `rust_future_poll` function inputs a callback function and an opaque pointer to call the callback with.
+   * If the future is pending, then the generated code registers a waker that will call the callback function with `RUST_FUTURE_MAYBE_READY`.
+     When the generated foreign code sees this, it calls poll again, starting the loop over.
+   * If the future is ready, then the callback function is immediately called with `RUST_FUTURE_READY` and we move to the next step.
+1. Call `rust_future_complete`, receiving the return value of the future
+1. Call `rust_future_free` (ideally in a `finally` block)
+1. Return the result from `rust_future_complete`
+
+### Why not have `rust_future_poll` return a boolean?
+
+As described above, `rust_future_poll` inputs a callback and an opaque pointer.
+If the future is ready, then `rust_future_poll` immediately invokes the callback, passing it the pointer back and `RUST_FUTURE_READY`
+A more straightforward API would be to return a boolean if the future is ready, why not do that?
+
+The issue is that it doesn't work well with the way foreign code's lifetime management.
+The opaque pointer is usually either something like an `Arc` or `Box` that holds data about the async call.
+When the foreign code passes the pointer to `rust_future_poll`, it temporarily leaks a reference (either an Arc refcount or the Box itself).
+Then when callback is called, it reconstructs the leaked reference using the pointer, undoing the leak and restoring balance to the universe.
+
+If `rust_future_poll` returned `true` to indicate that the future is ready, then the foreign code is in an awkward situation.
+It wants to use the async call data that the pointer was referring to, but it just sent a leaked pointer to Rust.
+It would have to hold on to a copy of the raw pointer, use that to restore the reference, and assume the callback is never going to be called.
```

Comment:
Raw pointer, opaque pointer, restore the reference.... There are a lot of terms. Could we have an index of terms and their definitions somewhere? The definition doesn't have to come from us!

Reply:
I love this idea, but I feel like that's out-of-scope for the async docs.  The best we currently have is this doc: https://mozilla.github.io/uniffi-rs/latest/internals/object_references.html. Maybe we could update that soon to also talk about foreign objects, future handles, etc.